README.md
## Synopsis

This tool allows you to list contents of all your accounts in a single page (characters, their stats and items, items in vaults). Also it generates a summary of all the items

## Difference between atomizer's muledump

* New
	- number 

## Download

All released versions are [here](https://github.com/HellRetard/muledump/releases).

## Howto

- unpack
- edit **`accounts_sample.js`**
- rename it to **`accounts.js`**
- open **`muledump.html`**

## Not so obvious features

- click on item to filter accounts that hold it
- _click on account name for individual options menu_ **does not work anymore**
- mouse over over account name for total left to max potions
- _ctrl-click account name to temporarily hide it from totals_ **does not work anymore**
- **as for `accounts.js` it's completely renewed**
	- removed row length, but added ordervault
	- ordervault=1 simply orders your vault like in realm, but for big vaults it might be uncomfortable to look at
	- ordervault=0 to disable vault order
- mouse over fame to display more info

## Upcomming features

- remove trading and prices from muledump completely
- 

CHANGELOG.md
**2017-08-13** version X.16.0-v2

- added alot -.-
- webhook test 2


**2016-07-25** version 0.6.5

- 27.7.DECA (1469395332) items
- combined stat display (base/max) by [JohnBlackburne](https://github.com/JohnBlackburne)

**2015-11-03** version 0.6.4

- [Kabam.com support](https://github.com/atomizer/muledump/wiki/Kabam-tutorial) (thanks to Zaltais333)
- consumable sorting fixes

**2015-09-22** version 0.6.3

- consumables are now sorted by type (thanks to Pfiffel & Nightfirecat)
- further improvements to the one-click login script (thanks to Nightfirecat)
- [Steam support](https://github.com/atomizer/muledump/wiki/Steam-Tutorial)
- a new item??? (27.7.0)

**2015-04-25** version 0.6.2

- one-click login now writes to both possible locations of the flash cookie
- fixed display of the one-click login links

**2015-02-22** version 0.6.1

- 27.3.1 items
- show all errors at the top of the page (as before)

**2015-01-27** version 0.6.0

- automated migration
- better user experience when mass-reloading
- hopefully will show more info in case of errors

**2015-01-18** version 0.5.1

- fixed cache not updating

**2015-01-17** version 0.5.0

- 27.3.0 items
- proper "migration required" warning
- automated ToS accept and age verification
- fixed PNG export (imgur)
- internal: moved stuff around, might break things
- internal: updated jquery

**2014-10-22** version 0.4.26

- 25.0 items

**2014-09-22** version 0.4.25

- 23.2 items

**2014-07-01** version 0.4.24

- 22.3 items
- one-click login script fix for latest version of AutoIt ([Nightfirecat](https://github.com/Nightfirecat))

**2014-06-09** version 0.4.22

- 21.4 items

really sorry for lack of updates recently :/


**2014-03-24** version 0.4.21

- 19.5 items

**2014-03-01** version 0.4.20

- gift chests option
- artemis skin

**2014-02-21** version 0.4.19

- 19.2 items

**2013-12-19** version 0.4.18

- 19.0 items

**2013-12-13** version 0.4.17

- 18.0 items, small ring positioning tweaks

**2013-09-14** version 0.4.16

- 16.4 items (in advance)
- gift chests ([Nightfirecat](https://github.com/Nightfirecat))
- more tolerant to corrupt char data (zub)
- disabled price lookup since the forums are dead
- email not verified warning

**2013-08-28** version 0.4.15

- 16.0 items
- 16.0 vault chest order ([Nightfirecat](https://github.com/Nightfirecat))
- display characters always in same order ([Nightfirecat](https://github.com/Nightfirecat))
- added `rowlength` variable to accounts.js
- better handling of unknown items (separate counters, item id in hex)
- use DarkDaemon's price guide ([aiedail92](https://github.com/aiedail92))
- updated JQuery and masonry
- more attempts to fix masonry-related rendering bugs

**2013-08-01** version 0.4.14

- 15.0 items
- hp/mp pot counters ([SlugKing](https://github.com/SlugKing), [Nightfirecat](https://github.com/Nightfirecat))
- feed power in tooltips
- copyright and license notes

**2013-03-31** version 0.4.13

- 12.1 items

**2013-02-08** version 0.4.12

- 11.0 items

**2013-01-22** version 0.4.11

- 9.0 items

**2012-12-12** version 0.4.10

- 7.0 items
- changed price guide to MustafaD (pull request by [avoxgames](https://github.com/avoxgames))

**2012-11-16** version 0.4.9

- 5.0 items
- added backpacks
- fixed ninja equips position in totals

**2012-11-2** version 0.4.8

- 4.0 items
- item names and tiers now match the in-game tooltips (except dosed items like elixirs)
- fixed char description sometimes spanning more lines than it should

**2012-10-25** version 0.4.7

- 3.1 items
- made achievement progress and additional stats options independent of each other

**2012-10-11** version 0.4.6

- ninja equipment fame bonuses
- fixed star color calculation

**2012-10-05** version 0.4.5

- ninja, 124.0 items

**2012-09-25** version 0.4.4

- new accounts.js option: "nomasonry" - set to 1 to turn off smart layout
- fixed some problems with per-account menus
- ctrl-click on name toggles the account in the totals

**2012-09-16** version 0.4.3

- authentic character portraits
- fixed long names breaking the box

**2012-09-04** version 0.4.1

- 123.5.0 items

**2012-08-25** version 0.4

- 123.4.4 items
- ability to set options for each account individually - click account name for menu (intended use: keep global options to items only and expand the good accounts as needed)
- added achievement progress calculation (thanks to Pfiffel, even though I didnt use your code)
- vault chests are now in their in-game order (thanks to Hals for assistance)
- improved page load time
- added redirect to Kable's video

**2012-08-05** version 0.3.2

- 123.4.1 items
- reduced probability of duplicate error messages

**2012-07-04** version 0.3.1

- 123.3.2 items
- fixed price guide url
- fixed some visual glitches

**2012-06-13** version 0.3: collaboration edition

- 123.3 stats
- export to TXT, CSV, JSON, PNG
- accuracy and god kill ratio
- fixed "sticky" options bug
- new reload symbol
- optional: one-click login -- by [FizzeBu](http://forums.wildshadow.com/user/24488); more info above
- optional: display prices from Kazansky -- by [aiedail92](https://github.com/aiedail92); more info above

**2012-05-08**

version 0.2.3

- fixed star counter
- better Opera compatibility
- various performance optimizations
- update checking is now on-demand

**2012-05-06** version 0.2.2

- stars
- automatic update checker

**2012-05-02** version 0.2.1

- fixed some visual bugs and edge cases introduced in previous update
- back to old method of issuing requests (slow but more stable)
- you can make options to stay open by clicking
- moved additional stats to the bottom

lib/totals.js
(function($, window) {

// weapons, armor, specials, rings, potions, unknown
var SLOT_ORDER = [1,2,3,17,8,24, 14,6,7, 4,5,11,12,13,15,16,18,19,20,21,22,23,25, 9, 10, 26, 100, 0];

var totals = {}, counters = {}, ids = [];

window.ids = ids

var options = window.options
var option_updated = window.option_updated
var items = window.items

function init_totals() {
	for (var id in items) {
		var itemname = items[id][0];
		if (itemname == 'Valentine Generator' || itemname == 'Beach Ball Generator' || itemname == 'Beer Slurp Generator') {
			// special generators (added to pets)
			items[id][1] = 26;
		} else if (items[id][1] == 10) {
			// set item tier for consumable sorting
			items[id][2] = consumable_tier(id, itemname);
		}
		ids.push(id);
	}
	// sort
	ids.sort(function(a, b) {
		a = items[a];
		b = items[b];
		function slotidx(it) { return SLOT_ORDER.indexOf(it[1]) }
		function tier(it) { return (it[2] < 0) ? 42 : it[2] }
		return (slotidx(a) - slotidx(b)) || (tier(a) - tier(b));
	});
}

function update_totals() {
	$('#totals').hide();
	var old = totals;
	totals = window.totals = {};
	var mules = window.mules

	function upd(arr) {
		if (!arr) return;
		var tokenCount;
		for (var i = 0; i < arr.length; i++) {
			var id = +arr[i];

			if (isbad(id)) continue;
			if (items[id][7] > 0)
				totals[id] = stackCount(id);
			// This is bad because if you have 3 stacks of the same token type
			// you will see 3 stacks in totals and they won't sum up together,
			// but honestly do we even need that? You are mostly about to stack
			// them up and then use it to get rewards from the tokens
			else if (id in totals) totals[id]++; else totals[id] = 1;
		}
	}

	function isbad(id) {
		if (!options.famefilter) return false;
		var i = items[id] || items[-1];
		if (i[5] > +options.fameamount) return false;
		return true;
	}

	// count items
	for (var i in mules) {
		var m = mules[i].items;
		if (mules[i].disabled || !m) continue;
		for (var j = 0; j < m.chars.length; j++) {
			var ch = m.chars[j];
			if (mules[i].opt('equipment')) upd(ch.slice(0, 4));
			if (mules[i].opt('inv')) upd(ch.slice(4, 12));
			if (mules[i].opt('backpack')) upd(ch.slice(12, 20));
		}
		if (mules[i].opt('vaults')) {
			for (j = 0; j < m.vaults.length; j++) {
				upd(m.vaults[j]);
			}
		}
	}

	for (i in old) {
		if (!items[i]) continue;
		if (!(i in totals)) {
			counters[i].hide();
			continue;
		}
		if (totals[i] != old[i]) {
			var a = totals[i];
			counters[i].find('div').toggle(a > 1).text(a);
		}
	}
	for (i in totals) {
		if (i in old) continue;
		if (!items[i]) continue
		if (!counters[i]) {
			var $i = window.item(i);
			var idx = ids.indexOf(i), minid = 0, minidx = 1e6, idxj = -1;
			for (var j in counters) {
				idxj = ids.indexOf(j);
				if (idxj > idx && idxj < minidx) {
					minidx = idxj;
					minid = j;
				}
			}
			if (minid) {
				$i.insertBefore(counters[minid]);
			} else $i.appendTo($('#totals'));
			counters[i] = $i;
		}
		var f = totals[i];
		counters[i].find('div').toggle(f > 1).text(f);
		counters[i].css('display', 'inline-block');
	}
	option_updated('totals');
}

// click-and-find

var filter = {};
window.filter = filter

function toggle_filter() {
	var $self = $(this);
	var id = $self.data('id');
	if (id in filter) delete filter[id]; else filter[id] = 1;
	window.relayout();
}

function update_filter() {
	$('.item.selected').filter(function() {
		return !($(this).data('id') in filter);
	}).removeClass('selected');
	$('.item').filter(function() {
		return $(this).data('id') in filter;
	}).addClass('selected');
	if ($.isEmptyObject(filter) || $('.item.selected:visible').length === 0) {
		var mules = window.mules
		for (var i in mules) if (mules[i].loaded) mules[i].dom.css('display', 'inline-block');
		return;
	}
	// if filtering
	var parts = ['equipment', 'inv', 'backpack', 'vaults', 'gifts'];
	$('.mule').each(function() {
		var $sel = $(this).find('.selected');
		for (var i = 0; i < $sel.length; i++) {
			var $par = $($sel[i]).parent();
			for (var j = 0; j < parts.length; j++) {
				var c = parts[j];
				if (options[c] && $par.hasClass(c)) {
					$(this).css('display', 'inline-block');
					return;
				}
			}
		}
		$(this).hide();
	});
}

function consumable_tier(id, itemname) {
	/*
	0: Vanity HP/MP
	1: HP/MP
	2: Stat Pots
	3: Wines
	4: Greater Stat Pots
	5: Incantation
	6: Dungeon Consumables
	7: Drake Eggs
	8: Tinctures and Effusions
	9: Holiday Candy and Alcohol
	10: Elixirs
	11: Shop Stuff
	12: Pet Food
	13: Keys
	14: Treasures
	15: Cards
	16: Skins
	17: Dyes
	18: Cloths
	*/
	switch (itemname) {
		case 'Minor Health Potion':
		case 'Minor Magic Potion':
		case 'Greater Health Potion':
		case 'Greater Magic Potion':
		case 'Potion of Health1':
		case 'Potion of Health2':
		case 'Potion of Health3':
		case 'Potion of Health4':
		case 'Potion of Health5':
		case 'Potion of Health6':
			return 0;
		case 'Health Potion':
		case 'Magic Potion':
			return 1;
		case 'Potion of Attack':
		case 'Potion of Defense':
		case 'Potion of Speed':
		case 'Potion of Vitality':
		case 'Potion of Wisdom':
		case 'Potion of Dexterity':
		case 'Potion of Life':
		case 'Potion of Mana':
			return 2;
		case 'Fire Water':
		case 'Cream Spirit':
		case 'Chardonnay':
		case 'Melon Liquer':
		case 'Cabernet':
		case 'Vintage Port':
		case 'Sauvignon Blanc':
		case 'Muscat':
		case 'Rice Wine':
		case 'Shiraz':
			return 3;
		case 'Greater Potion of Attack':
		case 'Greater Potion of Defense':
		case 'Greater Potion of Speed':
		case 'Greater Potion of Vitality':
		case 'Greater Potion of Wisdom':
		case 'Greater Potion of Dexterity':
		case 'Greater Potion of Life':
		case 'Greater Potion of Mana':
			return 4;
		case 'Wine Cellar Incantation':
			return 5;
		case 'Snake Oil':
		case 'Healing Ichor':
		case 'Pirate Rum':
		case 'Magic Mushroom':
		case 'Coral Juice':
		case 'Pollen Powder':
		case 'Holy Water':
		case 'Ghost Pirate Rum':
		case 'Speed Sprout':
			return 6;
		case 'Tincture of Fear':
		case 'Tincture of Courage':
		case 'Tincture of Dexterity':
		case 'Tincture of Life':
		case 'Tincture of Mana':
		case 'Tincture of Defense':
		case 'Effusion of Dexterity':
		case 'Effusion of Life':
		case 'Effusion of Mana':
		case 'Effusion of Defense':
			return 8;
		case 'Bahama Sunrise':
		case 'Blue Paradise':
		case 'Pink Passion Breeze':
		case 'Lime Jungle Bay':
		case 'Mad God Ale':
		case 'Oryx Stout':
		case 'Realm-wheat Hefeweizen':
		case 'Rock Candy':
		case 'Red Gumball':
		case 'Purple Gumball':
		case 'Blue Gumball':
		case 'Green Gumball':
		case 'Yellow Gumball':
		case 'Candy Corn':
			return 9;
		case 'Elixir of Health 7':
		case 'Elixir of Health 6':
		case 'Elixir of Health 5':
		case 'Elixir of Health 4':
		case 'Elixir of Health 3':
		case 'Elixir of Health 2':
		case 'Elixir of Health 1':
		case 'Elixir of Magic 7':
		case 'Elixir of Magic 6':
		case 'Elixir of Magic 5':
		case 'Elixir of Magic 4':
		case 'Elixir of Magic 3':
		case 'Elixir of Magic 2':
		case 'Elixir of Magic 1':
			return 10;
		case 'Small Firecracker':
		case 'Large Firecracker':
		case 'Sand Pail 5':
		case 'Sand Pail 4':
		case 'Sand Pail 3':
		case 'Sand Pail 2':
		case 'Sand Pail 1':
		case 'Transformation Potion':
		case 'XP Booster':
		case 'XP Booster Test':
		case 'Loot Tier Potion':
		case 'Loot Drop Potion':
		case 'XP Booster 1 hr':
		case 'XP Booster 20 min':
		case 'Backpack':
		case 'Old Firecracker':
		case 'Draconis Potion':
		case 'Lucky Clover':
		case 'Saint Patty\'s Brew':
			return 11;
		case 'Soft Drink':
		case 'Fries':
		case 'Great Taco':
		case 'Power Pizza':
		case 'Chocolate Cream Sandwich Cookie':
		case 'Grapes of Wrath':
		case 'Superburger':
		case 'Double Cheeseburger Deluxe':
		case 'Ambrosia':
			return 12;
		case 'Treasure Map':
			return 13;
		case 'Golden Ankh':
		case 'Eye of Osiris':
		case 'Pharaoh\'s Mask':
		case 'Golden Cockle':
		case 'Golden Conch':
		case 'Golden Horn Conch':
		case 'Golden Nut':
		case 'Golden Bolt':
		case 'Golden Femur':
		case 'Golden Ribcage':
		case 'Golden Skull':
		case 'Golden Candelabra':
		case 'Holy Cross':
		case 'Pearl Necklace':
		case 'Golden Chalice':
		case 'Ruby Gemstone':
			return 14;
		default:
			var lastword = itemname.split(" ");
			var itemtype = lastword[lastword.length - 1];
			if (itemtype == "Egg") {
				return 7;
			} else if(itemtype == "Key") {
				return 13;
			} else if(itemtype == "Card") {
				return 15;
			} else if(itemtype == "Skin") {
				return 16;
			} else if(itemtype == "Dye") {
				return 17;
			} else if(itemtype == "Cloth") {
				return 18;
			} else { // unknown item, abandon ship! (leave its tier at -1)
				console.log("Unknown item: " + id + " " + items[id][0]);
				return -1;
			}
		// end switch
	}
}


window.init_totals = init_totals
window.update_totals = update_totals
window.update_filter = update_filter
window.toggle_filter = toggle_filter

})($, window)


lib/functions.js
function stackCount(id){
	// Old tokens are yet to be added
	// Moss, Stone, Mad God Tokens
	if(id >= 0x99f && id <= 0x9a8) return id - 0x99f + 1;
	else if(id >= 0x9a9 && id <= 0x9b2) return id - 0x9a9 + 1;
	else if(id >= 0x9b3 && id <= 0x9bd) return id - 0x9b3 + 1;

	// Sand Pails
	else if(id >= 0xc34 && id < 0xc38) return 0xc38 - id + 1;

	// Elixirs
	else{
		switch(id){
			case 0xb18: return 7; break;
			case 0xb19: return 6; break;
			case 0xb1a: return 5; break;
			case 0xb1b: return 4; break;
			case 0xa4a: return 3; break;
			case 0xaed: return 2; break;
			case 0xaee: return 1; break;

			case 0xb14: return 7; break;
			case 0xb15: return 6; break;
			case 0xb16: return 5; break;
			case 0xb17: return 4; break;
			case 0xa81: return 3; break;
			case 0xa48: return 2; break;
			case 0xa49: return 1; break;
		}
	}
}


lib/mule.js
(function($, window) {

var options = window.options
var accounts = window.accounts
var items = window.items
var classes = window.classes

// max width of an account box in columns
var ROW = window.ordervault ? 15 : 7;

// dom snippet generators

function stat(where, type, text) {
	return $('<strong class="stat">').addClass(type).text(text).appendTo(where);
}

function item(id) {
	id = +id
	var ids = '0x' + id.toString(16)
	var $r = $('<div class="item">').data('id', id)

	var it = items[id]
	if (!it) {
		it = items[id] = ['item ' + ids, 0, -1, 0, 0, 0, 0]
	}
	if(it[7] > 0) { // is token ?
		var sc = stackCount(id);
		if(sc > 1)
			$r.append($('<div>').text(sc)); // is token ?
	}
	else $r.append($('<div>').text('0').hide());

	if (id != -1 && it[1] === 0) {
		$r.append($('<span>').text(ids))
	}
	
	var title = it[0];
	if (~it[2] && it[1] != 10 && it[1] != 9) title += ' (T' + it[2] + ')';
	if (it[6]) title += '\nFeed Power: ' + it[6]
	return $r.attr('title', title)
		.css('background-position', '-' + it[3] + 'px -' + it[4] + 'px')
}
window.item = item

function item_listing(arr, classname) {
	var $r = $('<div class="itemsc">');
	for (var i = 0; i < arr.length; i++) {
		item(arr[i]).appendTo($r);
	}
	if (classname) $r.addClass(classname);
	return $r;
}

function maketable(classname, items, row) {
	row = row || ROW;
	var $t = $('<table>').addClass(classname);
	var $row;
	for (var i = 0; i < items.length; i++) {
		if (i % row === 0) {
			if ($row) $t.append($row);
			$row = $('<tr>');
		}
		$('<td class="cont">').append(items[i]).appendTo($row);
	}
	if ($row) $t.append($row);
	var cols = items.length >= row ? row : items.length;
	cols = cols || 1;
	//$t.css('width', '' + (180 * cols + 14 * (cols - 1)) + 'px');
	return $t;
}

function addstar($t) {
	var $s = $('<span>').addClass('scont');
	$('<span>').text('666').appendTo($s);
	var $st = $('<span>').text('\u2605').addClass('star');
	$st.css('color', '#f07');
	$st.appendTo($s);
	$s.appendTo($t);
}

function addreloader(mule, target) {
	var rld = $('<div class="button">')
	rld.text('\u21bb')
	if (mule.data) {
		var updated = new Date(mule.data.query.created)
		rld.attr('title', 'last updated: ' + updated.toLocaleString())
	}
	rld.click(function(){ mule.reload() })
	rld.appendTo(target)
}

function mulelink(guid) {
	function toHex(s) {
		var r = '', t = '';
		for (var i = 0; i < s.length; i++) {
			t = s.charCodeAt(i).toString(16);
			if (t.length == 1) t = '0' + t;
			r += t;
		}
		return r;
	}
	var l = $('<a>').addClass('button');
	l.text('\u21d7');
	l.attr('href', 'muledump:' + toHex(guid) + '-' + toHex(accounts[guid]));
	l.attr('title', 'open this account');
	return l;
}

var ordervault = window.ordervault || 0;

var VAULTORDER = [
	-1,-1,-1,-1,-1,	26,21,17,22,27,	-1,-1,-1,-1,-1,
	-1,-1,-1,-1,-1,	14,12,11,13,15,	-1,-1,-1,-1,-1,
	-1,-1,-1,-1,-1,	 9,06,04,07,10,	-1,-1,-1,-1,-1,
	69,57,45,33,19,	05,02,01,03, 8,	23,35,47,60,73,
	67,55,43,29,16,	-1,-1,-1,-1,-1,	18,32,44,56,68,
	70,58,46,34,20,	-1,-1,-1,-1,-1,	24,36,48,59,74,
	75,63,49,37,25,	-1,-1,-1,-1,-1,	28,38,50,64,76,
	77,65,53,41,30,	-1,-1,-1,-1,-1,	31,42,54,66,78,
	79,71,61,51,39,	-1,-1,-1,-1,-1,	40,52,62,72,80
];

function arrangevaults(v) {
	var r = [], i;
	if(!ordervault){
		for (i = 0; i < v.length; i++) {
			r.push(v[i]);
		}
	}else{
		for (i = 0; i < VAULTORDER.length; i++) {
			r.push(v[ VAULTORDER[i] - 1 ]);
		}
	}
	return [ROW, r];
}

// Mule object

var Mule = function(guid) {
	if (!guid || !(guid in accounts)) return;
	this.guid = guid;
	this.fails = 0;
	this.dom = $('<div class="mule">');
	this.dom.appendTo($('#stage')).hide();
}

Mule.prototype.opt = function(name) {
	var o = options[this.guid];
	if (o && name in o) {
		return o[name];
	}
	return options[name];
}

Mule.prototype.cache_id = function() {
	return 'muledump:' + (!!window.testing ? 'testing:' : '') + this.guid
}

Mule.prototype.log = function(s, cl) {
	if (!this.overlay) {
		this.overlay = $('<div class="overlay">')
		var c = $('<div class="button">').text('X').appendTo(this.overlay)
		c.click(function() {
			$(this).parent().hide()
		})
		this.overlay.append($('<div class="log">'))
		this.overlay.appendTo(this.dom)
	}
	this.overlay.show()
	var log = this.overlay.find('.log')
	cl = cl || 'info'
	$('<div class="line">').text(s).addClass(cl).appendTo(log)
}

Mule.prototype.error = function(s) {
	this.log(s, 'error')
	var err = $('<div>')
	err.text(this.guid + ': ' + s)
	err.appendTo($('#errors'))
	addreloader(this, err)
	err.find('.button').click(function() { $(this).parent().remove() })
}

Mule.prototype.query = function(ignore_cache) {
	var self = this;
	if (this.busy) return; // somewhat protects against parallel reloads
	this.busy = true;
	this.loaded = false;
	$('#accopts').hide().data('guid', '');

	// read from cache if possible
	if (!ignore_cache) {
		var c = '';
		try {
			c = localStorage[this.cache_id()];
			c = JSON.parse(c);
		} catch(e) {}
		if (c) {
			this.parse(c);
			this.busy = false;
			return;
		}
	}

	var CR = { guid: this.guid }
	var pass = accounts[this.guid] || ''

	var platform = this.guid.split(':')[0]
	if (['kongregate', 'steamworks', 'kabam'].indexOf(platform) >= 0) {
		CR.secret = pass
	} else {
		CR.password = pass
	}

	this.log('loading data')
	window.realmAPI('char/list', CR, function(xhr) {
		xhr.done(onResponse).fail(onFail)
	})

	function onFail() {
		self.log('failed')
		self.busy = false;
		self.fails++;
		if (self.fails < 5) {
			self.query(true);
		} else {
			self.error('failed too many times, giving up');
		}
	}

	function onResponse(data) {
		self.busy = false;
		if (!data.query || !data.query.results) {
			self.error(data.query ? 'server error' : 'YQL service denied');
			self.log(data.query.results);
			if (data.query) {
				self.log('full response:' + JSON.stringify(data.query))
			}
			return;
		}
		var res = data.query.results

		function watchProgress(percent) {
			if (typeof percent != 'string') {
				self.error('migration failed')
				return
			}
			if (percent == '100') {
				self.reload()
				return
			}
			self.log('migration: ' + percent + '%')
			window.realmAPI('migrate/progress', CR, function(xhr) {
				xhr.fail(onFail).done(function(data) {
					var res = data && data.query && data.query.results
					var per = res.Progress && res.Progress.Percent
					watchProgress(per)
				})
			})
		}

		if (res.Migrate) {
			self.log('attempting migration')

			window.realmAPI('migrate/doMigration', CR, { iframe: true }, function() {
				watchProgress('0')
			})
			return
		}

		if (!res.Chars) {
			self.error(res.Error || 'bad reply: ' + JSON.stringify(res))
			return;
		}

		res = res.Chars

		if ('TOSPopup' in res) {
			window.realmAPI('account/acceptTOS', CR, { iframe: true })
		}

		if (res.Account && res.Account.IsAgeVerified != 1) {
			CR.isAgeVerified = 1
			window.realmAPI('account/verifyage', CR, { iframe: true })
		}

		self.parse(data)
	}
}

Mule.prototype.reload = function() {
	this.fails = 0
	if (this.overlay) this.overlay.find('.log').empty()
	this.query(true)
}

var PROPTAGS = 'ObjectType Level Exp CurrentFame'.split(' ')
var STATTAGS = 'MaxHitPoints MaxMagicPoints Attack Defense Speed Dexterity HpRegen MpRegen'.split(' ')
var STATABBR = 'HP MP ATT DEF SPD DEX VIT WIS'.split(' ')
var LTMTOTAL = [0,0,0,0,0,0,0,0,0,0];

Mule.prototype.parse = function(data) {
	if (this.overlay) this.overlay.hide()

	var d = data.query.results.Chars
	d = {
		Char: d.Char,
		Account: d.Account || {}
	}
	data.query.results.Chars = d

	// check if data changed?
/*	if (this.data && compare(d, this.data.query.results.Chars)) {
		return
	}*/

	this.data = data
	this.dom.hide().empty()
	this.overlay = null

	// write cache
	try {
		localStorage[this.cache_id()] = JSON.stringify(data);
	} catch(e) {}


	if (this.opt('guid')) {
		$('<input type="text" readonly="readonly">')
		.addClass('guid').val(this.guid).appendTo(this.dom);
		$('<br>').appendTo(this.dom);
	}

	addreloader(this, this.dom)

	if (!('VerifiedEmail' in d.Account)) {
		var $warn = $('<span class="button warn">').text('!!')
		$warn.attr('title', 'email not verified').appendTo(this.dom)
	}

	if (window.mulelogin) this.dom.append(mulelink(this.guid, accounts[this.guid]));

	d.Account = d.Account || {}
	var $name = $('<div>').addClass('name').text(d.Account.Name || '(no name)');

	var carr = [];
	if (d.Char) { // stupid array/object detection
		if (!d.Char.length) carr = [d.Char]; else carr = d.Char;
	}

	LTMTOTAL = [0,0,0,0,0,0,0,0,0,0];
	$totals = $('<div>').addClass('totals').text('Total pots left to max').append($('<hr>'));
	for (var i = 0; i < carr.length; i++) {
		var c = carr[i];
		if (!c) continue;
		var cl = classes[c.ObjectType];
		if (!cl) continue;

		for(var t = 0;t < STATTAGS.length; t++){	
			var s = +c[STATTAGS[t]] || 0;		
			var ltm = cl[3][t] - s;
			if (t < 2)
				ltm = Math.ceil(ltm / 5);
			LTMTOTAL[t] += ltm;
		}
	}
	for(var t = 0;t < STATTAGS.length; t++){
		$('<div>')
		.append($('<span>').addClass('sname').text(STATABBR[t] + ':'))
		.append($('<span>').addClass('stat max').text(LTMTOTAL[t])).appendTo($totals);
	}
	$name.append($totals);

	addstar(this.dom);
	$name.appendTo(this.dom);
	$('<br>').appendTo(this.dom);

	this.items = { chars: [], vaults: [] };

	var f = false;
	var arr = [];
	carr.sort(function(a,b) {return a.id - b.id});
	for (var i = 0; i < carr.length; i++) {
		var c = carr[i], $c = $('<div class="char">');
		if (!c) continue;
		var cl = classes[c.ObjectType];
		if (!cl) continue;
		if (this.opt('chdesc')) {
			f = true;
			var portimg = $('<img class="portrait">');
			var ratio = c.Exp/c.CurrentFame;
			var color = [
			(255 * (ratio/2000)).toFixed(0),
			(255 * (1-(ratio/2000))).toFixed(0),
			0];
			window.portrait(portimg, c.ObjectType, c.Tex1, c.Tex2);
			$('<div class="chdesc">')
				.append(portimg)
				.append($('<div class="fame">').text(c.CurrentFame + ' fame'))
				.append($('<div class="addons">'
					+ '<div class="level">' + ' ' + (c.Level < 20 ? 'level ' + c.Level + ' / ': '') + c.id + 'th</div>'
					+ '<div class="xp">' + c.Exp + ' XP </div>'
					+ '<div class="xp2" style="color: rgb('+color[0]+','+color[1]+','+color[2]+');">'+ ratio.toFixed(2) + ' XP per fame</div>'))
				.append($('<hr style="background: #aaa;">'))
				.appendTo($c);
		}
		if (this.opt('stats')) {
			f = true;
			var $stats = $('<table class="stats">');

			for (var t = 0; t < STATTAGS.length; t++) {
				var $row
				if (t % 2 === 0) $row = $('<tr>');
				$('<td class="sname">').text(STATABBR[t]).appendTo($row);
				var $s = $('<td>');
				var s = +c[STATTAGS[t]] || 0;
				var stt = this.opt('sttype');
				if (stt == 'base') {
					stat($s, 'base', s).toggleClass('maxed', s == cl[3][t]);
				} else if (stt == 'avg') {
					var avgd = s - Math.floor(cl[1][t] + (cl[2][t] - cl[1][t]) * (+c.Level - 1) / 19);
					stat($s, 'avg', (avgd > 0 ? '+' : '') + avgd)
						.addClass(avgd > 0 ? 'good' : (avgd < 0 ? 'bad' : ''))
						.toggleClass('very', Math.abs(avgd) > 14);
				} else if (stt == 'max') {
					var l2m = cl[3][t] - s;
					if (t < 2) l2m = l2m + ' (' + Math.ceil(l2m / 5) + ')';
					stat($s, 'max', l2m)
						.toggleClass('maxed', cl[3][t] <= s);
				} else if (stt == 'comb') {
					if (s < cl[3][t]) {
						var avgd = s - Math.floor(cl[1][t] + (cl[2][t] - cl[1][t]) * (+c.Level - 1) / 19);
						stat($s, 'small', s + ' / ' + cl[3][t])
							.addClass(avgd > 0 ? 'good' : (avgd < 0 ? 'bad' : ''))
							.toggleClass('very', Math.abs(avgd) > 14)
							.attr('title', avgd + ' from avg');
					} else {
						stat($s, 'maxed', s)
					}
				}
				$s.appendTo($row);
				if (t % 2) $row.appendTo($stats);
			}
			$c.append($stats);
		}

		// items
		var eq = (c.Equipment || '').split(',');
		this.items.chars.push(eq);
		var dobp = this.opt('backpack') && +c.HasBackpack
		if (this.opt('equipment') || this.opt('inv') || dobp) {
			f = true;
			var itc = $('<div>').addClass('items');
			if (this.opt('equipment')) itc.append(item_listing(eq.slice(0, 4), 'equipment'));
			if (this.opt('inv')) itc.append(item_listing(eq.slice(4, 12), 'inv'));
			if (dobp) itc.append(item_listing(eq.slice(12,20), 'backpack'));
			itc.appendTo($c);
		}
		if (this.opt('pcstats') || this.opt('goals')) {
			f = true;
			$c.append(window.printstats(c, d, this.opt('goals'), this.opt('pcstats')));
		}
		arr.push($c);
		$c.appendTo(this.dom);
	}
	if (f) {
		this.dom.append($('<hr class="chars">'));
	}
	arr = [];

	function makechest(items, classname) {
		var il = item_listing(items.slice(0, 8), classname)
		return $('<div class="items">').append(il)
	}

	if (this.opt('vaults')) {
		this.dom.append($('<hr class="vaults">'));
		// gift chest
		var gifts = d.Account.Gifts;
		if(gifts && this.opt('gifts')) {
			var items = gifts.split(',').reverse();
			this.items.vaults.push(items);  // for totals
			var garr = []
			while (items.length) {
				while (items.length < 8) items.push(-1)
				garr.push(makechest(items, 'gifts'))
				items = items.slice(8);
			}
			maketable('giftchest', garr, ROW).appendTo(this.dom)
		}

		// vault
		var chests = d.Account.Vault ? d.Account.Vault.Chest || ['-1'] : ['-1'];
		if (typeof chests == 'string') chests = [chests];
		var w = arrangevaults(chests);
		chests = w[1];

		for (i = 0; i < chests.length; i++) {

			if(ordervault){
				if(VAULTORDER[i] === -1){
					arr.push(null);
					continue;
				}else if(chests[i] === undefined){
					arr.push('<div class="locked"></div>');
					continue;
				}
			}

			var chest = (chests[i] || '-1').split(',');

			while (chest.length < 8)
				chest.push(-1);
			this.items.vaults.push(chest);
			arr.push(makechest(chest, 'vaults'));
		}
		maketable('vaults', arr, w[0]).appendTo(this.dom);
	}
	this.loaded = true;
	this.dom.css('display', 'inline-block')
	window.relayout();
}

window.Mule = Mule

})($, window)